### ISSUE TITLE
GPS accuracy circles permanently drawn on the map

### ISSUE THREAD

**Issue by gaellafond** [cgeo/cgeo#9552]
**Describe the bug:**
The GPS location have a circle around it to show its accuracy. When the map is panned or zoomed, the circle become permanently part of the map.

**To Reproduce:**
Steps to reproduce the behaviour:
1. Open the "Live map" in "c:geo"
2. Pan or zoom the map while the accuracy updates (while walking on the footpath for example)
3. Watch the trail of accuracy circle left behind.

NOTE:
- Quitting and reloading the map **do not** fix the problem.
- Switching to OpenStreetMap and back to Google map reset the map (get rid of the circles)
- "Force stop" the app and reload it also works

**Actual behaviour/state after performing these steps:**
GPS accuracy circles stay on the map.

**Expected behaviour/state after performing these steps:**
The old GPS accuracy circles should disappear as soon as a new circle is drawn.

**Version of c:geo used:**
You will find the c:geo version in c:geo Menu -> About c:geo
2020.12.03
NOTE: This bug have been happening for a long time. Could be related to the version of Android.

**Is the problem reproducible:**
Yes, it happens every time I use the app

**Screenshots:**
I created a trail of GPS accuracy circles by panning the map as I was walking around
![2020-12-16](https://user-images.githubusercontent.com/936244/102422443-c52c5180-4041-11eb-8f3d-189d7f1168a5.png)

**System information:**
```
--- System information ---
Device: Nokia 6.2 (Starlord_00WW, Nokia)
Android version: 10
Android build: 00WW_2_390
c:geo version: 2020.12.03
Google Play services: enabled - 20.47.13 (120400-344095733)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide caches: -
Hide waypoints: -
HW acceleration: enabled (default state)
System language: en_AU
System date format: d/M/yy
Debug mode active: no
System internal c:geo dir: /data/user/0/cgeo.geocaching (32.0 GB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (32.0 GB free) external non-removable
Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (32.0 GB free) external non-removable
Database: /data/user/0/cgeo.geocaching/databases/data (5.2 MB) on system internal storage
GPX import path: /storage/emulated/0/cgeo/gpx
GPX export path: /storage/emulated/0/cgeo/gpx
Offline maps path: null
Map render theme path:
Live map mode: true
Global filter: display all caches
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
geocaching.com: Logged in (Login OK) / BASIC
Geocaching.com date format: dd-MM-yyyy
Installed c:geo plugins: none
BRouter connection available: false
--- End of system information ---
```

**Additional context:**
N/A

**Comment by MagpieFourtyTwo** [comment]
Strange - never ever have seen s/th like this.
And I'm using c:geo since Android 4.x ... with different manufacturers (HTC, Sony, Huawei, but no Nokia).

BTW: Does anybody know why the proximity circle is not visible at all in OSM ...?

**Comment by gaellafond** [comment]
@MagpieFourtyTwo Thanks for getting to my issue so quickly. Just ask me if you want me to perform some tests.
I just learned that the circle is called "proximity circle", thanks. The bug only appear on Google maps, on my phone (both "Google: Map" and "Google: Satellite"). The bug occurs with both Google Maps. Switching between Google maps do not get rid of the circles. I need to switch to a OSM and back to Google to get rid of them.
The bug is inconsistent. Sometime, I can pan and zoom around without problem. Some other time, it will leave a trail of circle simply by panning around normally. It seems to appear more often when I zoom out a lot at once.

**Comment by gaellafond** [comment]
I have download a map for offline use. The issue do not occur with the downloaded map, but there is no proximity circle either.

**Comment by fm-sys** [comment]
I can reproduce the problem. Happens to me too. I thought I had already created an issue, but can't find. Probably this is a device specific problem?

I'm using Android 10 on a Huawei Phone...

**Comment by Lineflyer** [comment]
Oh...just tried it out. I can reproduce it on all my test devices. The problem is there as soon as I startup c:geo with Google Maps...not after switching to GMaps from other map types.
I am a bit puzzled why I did not see it earlier...probably I never started cgeo with GMaps beeing active.

**Comment by Lineflyer** [comment]
Correction: On Android 10, 11 but not on Android 5.

**Comment by MagpieFourtyTwo** [comment]
I am using GMaps just occasionally and never ever initially, too, and OSM doesn't provide the proximity circle, that's why I haven't seen it yet, neither ... however, proximity circle zombies do not occur in my c:geo - at least not while "normal" usage, even if I pinch and zoom like hell.

But: Was nevertheless able to reproduce, by using [Lockito](https://play.google.com/store/apps/details?id=fr.dvilleneuve.lockito&hl=de&gl=US) to perform one of my virtual test rides. ;)  It's fake start is about 10 km away from my real position, and as soon as I started the mock position, the position marker jumped away and indeed left a zombie circle at my real position. And subsequently it leaves even more zombies - but not on every display of a circle (i. e. position update), but on every 5th or up to 10th position.

Main difference to "normal" usage: This Lockito test ride is pretty quick, running at 240 km/h ;) - but at 5 km/h even with Lockito running for several minutes and dozens of zooming in and out not a single zombie occured ...

**Comment by MagpieFourtyTwo** [comment]
Yep, tested a few more, and the bigger the steps between two position updates are, the more likely it is, to get a zombie.
At least over here.

**Comment by MagpieFourtyTwo** [comment]
And all of this of course - like @Lineflyer already stated - only if I started c:geo with having set map to GM in a previous session.

**Comment by Lineflyer** [comment]
For me it already shows up just leaving the phone on my desk with the position jumping around a bit due to bad signal in a building. A lot of overlapping circles after some minutes.

**Comment by MagpieFourtyTwo** [comment]
Tried to do so, but most regretfully Huawei's GPS seems to be too good ... even inhouse very small circle and no jumps at all.
And the way to coffee machine and back is most obviously not long enough to produce zombies - or I am too slow ...  ;)

**Comment by moving-bits** [comment]
> And the way to coffee machine and back is most obviously not long enough to produce zombies - or I am too slow ... ;)

Maybe you need to get your coffee at 240 km/h ;-)

**Comment by gaellafond** [comment]
Thanks for your feedback. I will play around with the app this afternoon. I will start the live map with OSM and switch to GM once I started moving around to see if that fix the issue.

**Comment by gaellafond** [comment]
P.S. I don't know if walking at 240 km/h help to reproduce the issue. I don't think I will try that one :D

**Comment by MagpieFourtyTwo** [comment]
Those who think like this, however, certainly also get FTFs less often ...
Special situations require special measures!
Thus: Always keep trying!  ;)

> After a few minutes using the app, the ghost circles started to show up again.

How many are "a few"? I just get the zombies on initial start with GM. But to be honest, I just tried for three or four minutes (as GMap is cra... ahem, not really suitable for caching anyway, IMO.

> more likely to start when I keep turning the screen on and off

Depending on the duration between screen off and on, this may possibly be the same as a fresh start ...?

**Comment by gaellafond** [comment]
I use google map because I find the proximity circles useful (not the ghost ones of course). If that feature could be implemented in the other maps, I would probably start using them.

> After a few minutes using the app, the ghost circles started to show up again.
> How many are "a few"?

I would say about 5 minutes, but the screen was off pretty much the whole time. I switched to Google map, look how far I was, switched the screen off and turned it back on about 5 minutes later when I felt I was getting close to the GZ. They usually start showing up when I zoom or pan around, but maybe that's just a coincidence.

> this may possibly be the same as a fresh start

I'm not sure about that. I would think a fresh start would get rid of previous ghost circles. They tend to stay there until I switch to OSM map or force stop the app.

**Comment by MagpieFourtyTwo** [comment]
> I use google map because I find the proximity circles useful (not the ghost ones of course). If that feature could be implemented in the other maps, I would probably start using them.

Please be aware that the proximity circle is by no means exact but in fact just a wild guess (although an educated one, but still a guess).  It's in fact nice to have, but just that and not more.  You can esp. not rely on really being inside the circle.

>  5 minutes

Ok, will give it a try next time being out in the wild.

> fresh start

My thought behind this was, that c:geo caches a lot of data, to reuse it on subsequent starts (w/o being removed from memory completely). And to me it appears as if the circles are seen as a part of the map, as soon as a) the map page cache has been written and b) the next circle does not know anything anymore about the previously drawn circle, in order to first eliminate it ...?

**Comment by moving-bits** [comment]
I've had a look into it a couple of days ago, but couldn't find any obvious error at a glance. New circles only get drawn after having deleted the existing ones, so that part seems to be ok. The only thing I can imagine right now is that the handle to the layer gets lots somehow, so that a new circle layer gets created (although the previous one still exists), and new circles are drawn on top of the old one. But I haven't been able to reproduce this behaviour reliably in the emulator yet, which makes it hard to check possible solutions :-(

**Comment by Lineflyer** [comment]
On my Google Pixel its easily reproducible:
- Start c:geo (Google Maps was set as active map before)
- Open live map
- Center to position
- Zoom out a bit and in again
- Wait for some GPS accuracy/position fluctuations (in my house...bad reception)

That's how it looks like after one minute (street names blanked by me):
![Screenshot_20201221-174933](https://user-images.githubusercontent.com/949669/102801295-50e10d80-43b5-11eb-93bb-429293250f28.png)

**Comment by MagpieFourtyTwo** [comment]
@moving-bits I guess you can create mock positions in the emulator?
If so, try to let it jump from current to a position which is several kilometers away.
And make sure that you've started c:geo with GMap already active.

Or can you use Lockito in the emulator ...?

**Comment by fm-sys** [comment]
On a real device GPS accuracy and position are charging frequently which is hard to reproduce in an emulator. But the emulator is a a full featured virtual Android device so installing and using lockito should be possible...

**Comment by moving-bits** [comment]
> @moving-bits I guess you can create mock positions in the emulator?
> If so, try to let it jump from current to a position which is several kilometers away.
> And make sure that you've started c:geo with GMap already active.

I've tried this in several variations, but except for the very first time, I got no second accuracy circle. And that single event did not happen ever again.

Also I tried using a physical device, but GPS signal is too weak to come through the shutters - I'm stuck in Paris :-(

**Comment by moving-bits** [comment]
This is really strange... I got it reproduced, finally, by using a physical device and Lockito app. So I started adding debug info to the code, namely the part where the accuracy circle and heading bitmap get removed and redrawn. Although I can see the effect of "forgotten circles" (why is it always circles, never the heading indicator?) the log looks ok to me:

```
12-21 21:53:26.400 20420 20536 D mali_winsys: new_window_surface returns 0x3000,  [1080x1920]-format:1
12-21 21:53:26.450 20420 20420 E cgeo    : [main] create GooglePositionAndHistory
12-21 21:53:26.460 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:26.460 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:26.470 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:26.470 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@d8e582e
12-21 21:53:26.470 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@d65ed5c
12-21 21:53:26.510 20420 20519 D SensorManager: registerListener :: 6, MPL Rotation Vector, 200000, 0,  
12-21 21:53:26.510 20420 20961 I qtaguid : Tagging socket 54 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:26.560 20420 20420 D ViewRootImpl: MSG_RESIZED_REPORT: ci=Rect(0, 72 - 0, 0) vi=Rect(0, 72 - 0, 0) or=1
12-21 21:53:26.570 20420 20995 D libEGL  : eglInitialize EGLDisplay = 0xd6fbf564
12-21 21:53:26.570 20420 20420 W cgeo    : [main] No center, cannot draw scale
12-21 21:53:26.570 20420 20995 D mali_winsys: new_window_surface returns 0x3000,  [1080x1680]-format:2
12-21 21:53:26.610 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:26.610 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@1a2f367
12-21 21:53:26.610 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@1353514
12-21 21:53:26.610 20420 20420 W cgeo    : [main] No center, cannot draw scale
12-21 21:53:26.640 20420 20961 I qtaguid : Untagging socket 80
12-21 21:53:26.650 20420 20420 I Timeline: Timeline: Activity_idle id: android.os.BinderProxy@8cbb394 time:5331388
12-21 21:53:26.670 20420 20962 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:26.670 20420 20962 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:26.670 20420 20962 I qtaguid : Tagging socket 80 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:26.740 20420 20962 I qtaguid : Untagging socket 80
12-21 21:53:26.790 20420 20963 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:26.790 20420 20963 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:26.790 20420 20963 I qtaguid : Tagging socket 80 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:26.860 20420 20963 I qtaguid : Untagging socket 80
12-21 21:53:26.900 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:26.900 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@f879959
12-21 21:53:26.900 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@14b991e
12-21 21:53:26.930 20420 20536 D libEGL  : eglInitialize EGLDisplay = 0xde27f614
12-21 21:53:26.930 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:26.930 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@cf180a6
12-21 21:53:26.930 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@e6a8ae7
12-21 21:53:26.950 20420 20420 V ActivityThread: updateVisibility : ActivityRecord{78b839 token=android.os.BinderProxy@b741c01 {cgeo.geocaching/cgeo.geocaching.MainActivity}} show : false
12-21 21:53:26.970 20420 20536 D libEGL  : eglInitialize EGLDisplay = 0xde27f614
12-21 21:53:27.020 20420 20536 D libEGL  : eglInitialize EGLDisplay = 0xde27f614
12-21 21:53:27.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:27.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@55c3571
12-21 21:53:27.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@6f15f56
12-21 21:53:27.410 20420 20964 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:27.410 20420 20964 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:27.460 20420 20964 I qtaguid : Tagging socket 48 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:28.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:28.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@70105c
12-21 21:53:28.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@7db2f65
12-21 21:53:28.360 20420 20996 W DynamiteModule: Local module descriptor class for com.google.android.gms.googlecertificates not found.
12-21 21:53:28.360 20420 20996 I DynamiteModule: Considering local module com.google.android.gms.googlecertificates:0 and remote module com.google.android.gms.googlecertificates:6
12-21 21:53:28.360 20420 20996 I DynamiteModule: Selected remote version of com.google.android.gms.googlecertificates, version >= 6
12-21 21:53:28.370 20420 20996 W System  : ClassLoader referenced unknown path: /data/data/com.google.android.gms/app_chimera/m/00000200/n/armeabi-v7a
12-21 21:53:28.380 20420 20996 W System  : ClassLoader referenced unknown path: /data/data/com.google.android.gms/app_chimera/m/00000200/n/armeabi
12-21 21:53:28.380 20420 20996 D ResourcesManager: For user 0 new overlays fetched Null
12-21 21:53:28.450 20420 20446 D LeakCanary: Setting up flushing for Thread[GoogleApiHandler,5,main]
12-21 21:53:28.840 20420 20964 I qtaguid : Tagging socket 97 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:29.100 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:29.100 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@78354c0
12-21 21:53:29.100 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@33dc5f9
12-21 21:53:29.270 20420 20964 I qtaguid : Untagging socket 48
12-21 21:53:30.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:30.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@67d3b31
12-21 21:53:30.090 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@4415216
12-21 21:53:30.390 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:30.390 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@1506cc6
12-21 21:53:30.400 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@786d487
12-21 21:53:30.400 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:30.400 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@93cfb4
12-21 21:53:30.400 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@17077dd
12-21 21:53:31.110 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:31.110 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@5739376
12-21 21:53:31.110 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@c67ed77
12-21 21:53:31.460 20420 20961 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:31.460 20420 20961 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:31.460 20420 20961 I qtaguid : Tagging socket 80 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:31.530 20420 20961 I qtaguid : Untagging socket 80
12-21 21:53:32.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:32.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@b797357
12-21 21:53:32.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@24ce44
12-21 21:53:32.370 20420 20962 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:32.370 20420 20962 I System.out: (HTTPLog)-Static: isSBSettingEnabled false
12-21 21:53:32.370 20420 20962 I qtaguid : Tagging socket 80 with tag 3000110100000000{805310721,0} uid -1, pid: 20420, getuid(): 10066
12-21 21:53:32.430 20420 20962 I qtaguid : Untagging socket 80
12-21 21:53:33.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:33.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@3b9d1ba
12-21 21:53:33.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@3e1286b
12-21 21:53:34.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:34.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@a8cefc2
12-21 21:53:34.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@11b61d3
12-21 21:53:34.380 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:34.380 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@55065c3
12-21 21:53:34.380 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@a0eee40
12-21 21:53:34.380 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:34.380 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@a732179
12-21 21:53:34.390 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@3ce4ebe
12-21 21:53:35.090 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:35.090 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@317dced
12-21 21:53:35.090 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@2c77722
12-21 21:53:36.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:36.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@a0f1915
12-21 21:53:36.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@d50442a
12-21 21:53:37.100 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:37.100 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@e192e82
12-21 21:53:37.100 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@7709993
12-21 21:53:38.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:38.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@c7d1f8a
12-21 21:53:38.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@f9ea9fb
12-21 21:53:38.390 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:38.390 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@96a07eb
12-21 21:53:38.390 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@b512948
12-21 21:53:38.390 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:38.390 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@1d2b3e1
12-21 21:53:38.390 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@a8b5506
12-21 21:53:39.110 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:39.110 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@28e62d5
12-21 21:53:39.110 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@e7b2aea
12-21 21:53:40.080 20420 20420 E cgeo    : [main] removeAll positionObj
12-21 21:53:40.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.CircleOptions@fe46af2
12-21 21:53:40.080 20420 20420 E cgeo    : [main] add positionObj: com.google.android.gms.maps.model.MarkerOptions@6037543
```

AFAICS each `add positionObj` is preceded by a `removeAll positionObj`, so all old circles should be cleared before drawing the new one. (BTW: Debug code is added to maps/google/v2/ `GoogleMapObjects` and `GooglePositionAndHistory`)

I'll investigate further... Currently I'm deep into the `GoogleMapObjectsQueue`, where add and remove do not sum up anymore - strange!

**Comment by moving-bits** [comment]
Ok, looks like I got the culprit:

Sometimes the same object (circle) get added twice to the drawing queue, probably due to timing issues. Such circles are both drawn onto the map's surface.
Before drawing, the `GoogleMapObjectsQueue` remember which objects have been drawn to be able to remove them later. Internally they are stored as a hash map - and that's where the problem is buried, because adding the same object twice to a hash map only stores it once, and so it can only be removed once, whereas Google Maps has two objects, leading to one object getting forgotten...

See PR #9565.

**Comment by gaellafond** [comment]
@moving-bits Nice found. Does the object collection needs to be a hashmap? Changing it to a list could solve the issue (both circle would get removed) but that could slow down the app.
Maybe we can add a check to see if the object is already in the hashmap before drawing it? Avoiding the duplicate circle seems like a better solution (it's better to prevent the problem for happening in the first place).

Timing issues are hard to reproduce. Maybe zooming or panning the map holds the drawing process which changes the timing?

I tried Lineflyer idea. I started the live map on GM and zoomed to my location, and let it run for about 5 minutes. I got the initial ghost circle, but no other one after that. Then I switch to another app and came back to c:geo and managed to get another ghost circle. Then I zoomed and panned around for 30 seconds and got about 10 more ghost circles during that 30 seconds.

One thing I discovered while doing this. Changing the orientation of the phone (portrait / landscape) refresh the map and get rid of ghost circles.

**Comment by moving-bits** [comment]
> @moving-bits Nice found. Does the object collection needs to be a hashmap? Changing it to a list could solve the issue (both circle would get removed) but that could slow down the app.

I guess it's a hashmap for speed reasons. As this function can be called several times per second and objects need to be compared, a hashmap seems like a reasonable idea.

> Maybe we can add a check to see if the object is already in the hashmap before drawing it?

That's what I'm doing in my PR. I had also tried the other way around - reference counting to be able to remove all objects, but the code not only got more complicated, there have been no visible advantage of this solution.

> Timing issues are hard to reproduce. Maybe zooming or panning the map holds the drawing process which changes the timing?

I was able to reproduce the issue only on a physical device and with the help of the app "Lockito", which simulates a contiuos movement along a user-defined trail. I created a trail being several km long and started the app in loop mode. After some time after starting c:geo with Google Maps the faulty circles started to appear.

**Comment by MagpieFourtyTwo** [comment]
This should probably be re-opened again, as the proximity circle (although perfectly fine working in OSM now) now gets deleted way too often in GMap: After zooming it shortly appears, then disappears again ...

**Comment by fm-sys** [comment]
Let's create a new issue for it...

**Comment by moving-bits** [comment]
Needed to reopen this issue, as the fix in #9565 has unwanted side-effects (see GM-related part of #9374).

**Comment by gaellafond** [comment]
I came back to the issue to confirm it was fixed in the latest version, to see that it was re-opened. Hopefully there will be a new fix instead of a rollback...

**Comment by gaellafond** [comment]
I noticed that the accuracy circle will sometime disappear while zooming or panning. But that's not really an issue, it reappears after a few seconds. I used the latest version of the app for a while and could not reproduce the "ghost" circles. Good work everyone! Thanks for looking into my issue.

**Comment by moving-bits** [comment]
Yeah, there's still some bug regarding the accuracy circles on Google Maps, which leads to them sometimes disappearing. The "fix" for the original issue is actually just a mitigation (preventing the circle getting drawn multiple times, but removed only once).

Anyway, thanks for testing and for your feedback!
